pyro/nn/module.py

```diff
@@ -369,14 +369,14 @@ def __setattr__(self, name, value):
             self._pyro_params[name] = constraint, event_dim
             if self._pyro_context.active:
                 fullname = self._pyro_get_fullname(name)
-                if fullname in _PYRO_PARAM_STORE:
-                    # Update PyroModule <--- ParamStore.
-                    unconstrained_value = _PYRO_PARAM_STORE._params[fullname]
-                    if not isinstance(unconstrained_value, torch.nn.Parameter):
-                        # Update PyroModule ---> ParamStore (type only; data is preserved).
-                        unconstrained_value = torch.nn.Parameter(unconstrained_value)
-                        _PYRO_PARAM_STORE._params[fullname] = unconstrained_value
-                        _PYRO_PARAM_STORE._param_to_name[unconstrained_value] = fullname
+                constrained_value = pyro.param(fullname, constrained_value,
+                                               constraint=constraint, event_dim=event_dim)
+                unconstrained_value = constrained_value.unconstrained()
+                if not isinstance(unconstrained_value, torch.nn.Parameter):
+                    # Update PyroModule ---> ParamStore (type only; data is preserved).
+                    unconstrained_value = torch.nn.Parameter(unconstrained_value)
+                    _PYRO_PARAM_STORE._params[fullname] = unconstrained_value
+                    _PYRO_PARAM_STORE._param_to_name[unconstrained_value] = fullname
```

Comment:
@fehiepsi could you please review the `PyroModule` change here? The purpose is to make `PyroModule.__setattr__()` effectful, i.e. to respect effect handlers. Otherwise the behavior should be unchanged.

Reply:
This looks correct to me.